src/mpegdemux.1
.TH MPEGDEMUX 1 "2009-02-13" "HH" "mpegdemux"

.SH NAME
mpegdemux \- MPEG1/2 system stream demultiplexer

.SH SYNOPSIS
.BI mpegdemux " [options] [input [output]]"

.SH DESCRIPTION
.BR mpegdemux (1)
is an MPEG1/MPEG2 system stream demultiplexer. It can be
used to list the contents of an MPEG system stream and to extract
elementary streams.

.BR mpegdemux (1)
has four primary modes of operation:
.TP
scan
In this mode the MPEG system stream is scanned for elementary
streams.
.TP
list
In this mode the contents of an MPEG system stream are listed
in a textual form. This is useful to get an overview of what's in
an MPEG file
.TP
demux
In this mode elementary streams are extracted from an MPEG
system stream. The system stream packet structure is dissolved in
the process. Typically each extracted stream is written to its
own file.
.TP
remux
This is like demux, except that the MPEG system stream
structure is left intact. This means that the output is again
an MPEG system stream with all but the selected elementary
streams removed.

.SH OPTIONS
.TP
.B -a, --ac3
AC3 sound packets in DVD MPEG2 streams have a 3 byte header
that is neither part of the MPEG specification nor of the
AC3 specification. When this option is used, these 3 bytes
are removed to produce a correct AC3 stream. Note that this
option applies to all selected substreams without checking
whether they actually contain an AC3 elementary stream.
\
.TP
.BI "-b, --base-name " name
When demultiplexing more than one stream, the output file names
can be set using this option. To generate the output file name
for a stream, every # character in \fIname\fR is replaced by
a hex digit of the stream id. For example, to extract all
video streams in one go, use something like

$ mpegdemux -d -b video_##.m1v -s 0xc0-0xcf src.mpg

to get files video_c0.m1v, video_c1.m1v, ...
\
.TP
.B -c --scan
Scan a system stream for elementary streams. This is the default
mode. All streams and substreams are automatically selected when
using this option.
\
.TP
.B -d, --demux
Demultiplex an MPEG system stream. The demultiplexed streams are
written to the output file unless the \fB--base-name\fR option is
used. If the \fB--base-name\fR option is not used, only one stream
can be demultiplexed (if more streams are specified, they will be
randomly interleaved in the output file).
\
.TP
.B -D, --no-drop
Don't drop incomplete packets in demuxing and remuxing mode.
\
.TP
.B -e, --no-end
Don't print end codes in listing mode. Additionally, in
remuxing mode, make sure that there is exactly one end
code at the end of the stream.
\
.TP
.B -E, --empty-packs
When streams are removed during remuxing, packs can become
empty. Including these empty packs in the output is pointless
and therefore is not done by default. Use this option to
force inclusion of all packs.
\
.TP
.B -F, --first-pts
In scan mode, in addition to each stream's first packet,
also list the packet with the lowest presentation time stamp.
\
.TP
.B -h, --no-system-headers
Don't print system headers in listing mode. Additionally, in
remuxing mode, don't repeat system headers.
\
.TP
.BI "-i, --invalid " spec
Select invalid streams. Packets of invalid streams are not recognized
as packets and their contents are parsed as MPEG system stream data
rather than being skipped. This is useful for broken/incomplete
streams. The syntax for \fIspec\fR is the same as for \fB-s\fR.
Additionally, if \fIspec\fR is "-" then all streams that have not yet
been selected by \fB-s\fR are made invalid.
\
.TP
.B -k, --no-packs
Don't print packs in listing mode.
\
.TP
.B "-K, --remux-skipped"
Copy bytes that are skipped while looking for a start code.
\
.TP
.B -l, --list
List the system headers, packs and packets in an MPEG system stream.
\
.TP
.BI "-m, --packet-max-size " size
Set the maximum packet size to \fIsize\fR. Packets in the input stream
that are larger are considered invalid. As with the \fB-i\fR option,
the packet is not simply skipped but parsed as MPEG system stream data.
\
.TP
.BI "-p, --substream " spec
This option selects private substreams. Whenever Private Stream 1
(0xbd) is selected using \fB-s\fR, the substreams within that private
stream can be selected using \fB-p\fR. The syntax for \fIspec\fR is
the same as for \fB-s\fR.
\
.TP
.BI "-P, --substream-map " "id1 id2"
Remap substream \fIid1\fR to \fIid2\fR when remuxing.
\
.TP
.B -r, --remux
Remultiplex an MPEG system stream. The output MPEG system stream
is written to the output file. Many options control what is
copied from the input to the output and what is discarded.
\
.TP
.BI "-s, --streams " spec
This option selects streams. \fIspec\fR specifies the stream IDs in
the following form:

[+|-]<id>[-<id>]{/[+|-]<id>[-<id>]}

where id is either a numeric stream ID or one of \fBall\fR or
\fBnone\fR. A "-" in front of an ID range means exclusion. For example
the spec

-s 0xc0-0xcf/-0xc2

selects all video streams (0xc0 - 0xcf) except stream 0xc2.
\
.TP
.BI "-S, --stream-map " "id1 id2"
Remap stream \fIid1\fR to \fIid2\fR when remuxing.
\
.TP
.B -t, --no-packets
Don't print packets in listing mode.
\
.TP
.B -u, --spu
This option is used to extract DVD subtitles. It is necessary
because the subtitle streams on DVD don't contain all the
timing information (the time stamps in the packet headers
are required). If this option is used during demultiplexing,
the output files for all substreams are written in the
following format:

"SPU " (4 bytes)

PTS (8 bytes, MSB first)
.br
SPU packet

PTS (8 bytes)
.br
SPU packet

and so on
\
.TP
.B -x, --split
Split the remuxed stream at sequence boundaries. This option is
only meaningful in remuxing mode. It can not be used together
with the \fB-e\fR option. The individual sequences are written to
files whose name was set with the \fB-b\fR option.
\
.TP
.B --help
Print usage information
\
.TP
.B --version
Print version information

.SH EXAMPLES

Scan a system stream for elementary streams:

$ mpegdemux -c -v src.mpg


List the contents of an MPEG system stream:

$ mpegdemux -l -k -s all -p all src.mpg


Extract the first video stream:

$ mpegdemux -d -s 0xe0 src.mpg dst.m1v


Extract all audio streams:

$ mpegdemux -d -s 0xc0-0xdf -b audio_##.mpa src.mpg


Remove the second video stream:

$ mpegdemux -r -s all/-0xc1 -p all src.mpg dst.mpg


Extract the first AC3 audio stream from a DVD MPEG2 system stream:

$ mpegdemux -d -s 0xbd -p 0x80 --ac3 src.mpg dst.ac3


Exchange the first and the second audio stream:

$ mpegdemux -r -s all -p all -S 0xc0 0xc1 -S 0xc1 0xc0 src.mpg dst.mpg

.SH SEE ALSO
.BR mplex "(1)"

.SH AUTHOR
Hampa Hug <hampa@hampa.ch>


NEWS
0.1.3	2009-02-15	r106

	* A packet size limit can now be specified to avoid losing large
	  parts of broken streams.
	* Mpegdemux can now remap stream IDs.
	* In remuxing mode skipped bytes can now be copied to the new
	  stream.

0.1.2	2004-09-19	r81

	* Streams can now be declared invalid
	* Implemented splitting system stream at end codes when remuxing
	* Made the transition from CVS to SVN
	* Bugfixes

0.1.1	2003-08-10	r61

	* Dummy release

0.1.0	2003-08-10	r60

	* Bugfixes
	* Private substreams can now be demultiplexed into separate files

0.0.4	2003-04-09	r46

	* Bugfixes
	* Better handling of skipped bytes
	* Incomplete packets can now be dropped

0.0.3	2003-03-08	r33

	* New scan mode
	* System stream syntax bug fix
	* Some changes to the command line parametrs

0.0.2	2003-03-06	r28

	* Lots of changes

0.0.1	2003-02-02	r1

	* First version


src/mpegdemux.c
/*****************************************************************************
 * mpegdemux                                                                 *
 *****************************************************************************/

/*****************************************************************************
 * File name:   src/mpegdemux.c                                              *
 * Created:     2003-02-01 by Hampa Hug <hampa@hampa.ch>                     *
 * Copyright:   (C) 2003-2009 Hampa Hug <hampa@hampa.ch>                     *
 *****************************************************************************/

/*****************************************************************************
 * This program is free software. You can redistribute it and / or modify it *
 * under the terms of the GNU General Public License version 2 as  published *
 * by the Free Software Foundation.                                          *
 *                                                                           *
 * This program is distributed in the hope  that  it  will  be  useful,  but *
 * WITHOUT  ANY   WARRANTY,   without   even   the   implied   warranty   of *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU  General *
 * Public License for more details.                                          *
 *****************************************************************************/


#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>

#include "message.h"
#include "mpeg_parse.h"
#include "mpeg_list.h"
#include "mpeg_demux.h"
#include "mpeg_remux.h"
#include "mpeg_scan.h"
#include "mpegdemux.h"


static unsigned par_mode = PAR_MODE_SCAN;

static FILE     *par_inp = NULL;
static FILE     *par_out = NULL;

unsigned char   par_stream[256];
unsigned char   par_substream[256];

unsigned char   par_stream_map[256];
unsigned char   par_substream_map[256];

int             par_no_shdr = 0;
int             par_no_pack = 0;
int             par_no_packet = 0;
int             par_no_end = 0;
int             par_empty_pack = 0;
int             par_remux_skipped = 0;
int             par_split = 0;
int             par_drop = 1;
int             par_scan = 0;
int             par_first_pts = 0;
int             par_dvdac3 = 0;
int             par_dvdsub = 0;
char            *par_demux_name = NULL;
unsigned        par_packet_max = 0;


static
void prt_help (void)
{
	fputs (
		"usage: mpegdemux [options] [input [output]]\n"
		"  -a, --ac3                    Assume DVD AC3 headers in private streams\n"
		"  -b, --base-name name         Set the base name for demuxed streams\n"
		"  -c, --scan                   Scan the stream [default]\n"
		"  -d, --demux                  Demultiplex streams\n"
		"  -D, --no-drop                Don't drop incomplete packets\n"
		"  -e, --no-end                 Don't list end codes [no]\n"
		"  -E, --empty-packs            Remux empty packs [no]\n"
		"  -F, --first-pts              Print packet with lowest PTS [no]\n"
		"  -h, --no-system-headers      Don't list system headers\n"
		"  -i, --invalid id             Select invalid streams [none]\n"
		"  -k, --no-packs               Don't list packs\n"
		"  -K, --remux-skipped          Copy skipped bytes when remuxing [no]\n"
		"  -l, --list                   List the stream contents\n"
		"  -m, --packet-max-size int    Set the maximum packet size [0]\n"
		"  -p, --substream id           Select substreams [none]\n"
		"  -P, --substream-map id1 id2  Remap substream id1 to id2\n"
		"  -r, --remux                  Copy modified input to output\n"
		"  -s, --stream id              Select streams [none]\n"
		"  -S, --stream-map id1 id2     Remap stream id1 to id2\n"
		"  -t, --no-packets             Don't list packets\n"
		"  -u, --spu                    Assume DVD subtitles in private streams\n"
		"  -x, --split                  Split sequences while remuxing [no]\n",
		stdout
	);
}

static
void prt_version (void)
{
	fputs (
		"mpegdemux version " MPEGDEMUX_VERSION_STR
		"\n\n"
		"Copyright (C) 2003-2009 Hampa Hug <hampa@hampa.ch>\n",
		stdout
	);
}

static
char *str_clone (const char *str)
{
	char *ret;

	ret = malloc (strlen (str) + 1);
	if (ret == NULL) {
		return (NULL);
	}

	strcpy (ret, str);

	return (ret);
}

static
int str_isarg (const char *str, const char *arg1, const char *arg2)
{
	if ((arg1 != NULL) && (strcmp (str, arg1) == 0)) {
		return (1);
	}

	if ((arg2 != NULL) && (strcmp (str, arg2) == 0)) {
		return (1);
	}

	return (0);
}

static
const char *str_skip_white (const char *str)
{
	while ((*str == ' ') || (*str == '\t')) {
		str += 1;
	}

	return (str);
}

static
int str_get_streams (const char *str, unsigned char stm[256], unsigned msk)
{
	unsigned i;
	int      incl;
	char     *tmp;
	unsigned stm1, stm2;

	incl = 1;

	while (*str != 0) {
		str = str_skip_white (str);

		if (*str == '+') {
			str += 1;
			incl = 1;
		}
		else if (*str == '-') {
			str += 1;
			incl = 0;
		}
		else {
			incl = 1;
		}

		if (strncmp (str, "all", 3) == 0) {
			str += 3;
			stm1 = 0;
			stm2 = 255;
		}
		else if (strncmp (str, "none", 4) == 0) {
			str += 4;
			stm1 = 0;
			stm2 = 255;
			incl = !incl;
		}
		else {
			stm1 = (unsigned) strtoul (str, &tmp, 0);
			if (tmp == str) {
				return (1);
			}

			str = tmp;

			if (*str == '-') {
				str += 1;
				stm2 = (unsigned) strtoul (str, &tmp, 0);
				if (tmp == str) {
					return (1);
				}
				str = tmp;
			}
			else {
				stm2 = stm1;
			}
		}

		if (incl) {
			for (i = stm1; i <= stm2; i++) {
				stm[i] |= msk;
			}
		}
		else {
			for (i = stm1; i <= stm2; i++) {
				stm[i] &= ~msk;
			}
		}

		str = str_skip_white (str);

		if (*str == '/') {
			str += 1;
		}
	}

	return (0);
}

char *mpeg_get_name (const char *base, unsigned sid)
{
	unsigned n;
	unsigned dig;
	char     *ret;

	if (base == NULL) {
		base = "stream_##.dat";
	}

	n = 0;
	while (base[n] != 0) {
		n += 1;
	}

	n += 1;

	ret = (char *) malloc (n);
	if (ret == NULL) {
		return (NULL);
	}

	while (n > 0) {
		n -= 1;
		ret[n] = base[n];

		if (ret[n] == '#') {
			dig = sid % 16;
			sid = sid / 16;
			if (dig < 10) {
				ret[n] = '0' + dig;
			}
			else {
				ret[n] = 'a' + dig - 10;
			}
		}
	}

	return (ret);
}

int mpeg_stream_excl (unsigned char sid, unsigned char ssid)
{
	if ((par_stream[sid] & PAR_STREAM_SELECT) == 0) {
		return (1);
	}

	if (sid == 0xbd) {
		if ((par_substream[ssid] & PAR_STREAM_SELECT) == 0) {
			return (1);
		}
	}

	return (0);
}

/* check if packet is valid. returns 0 if it is. */
int mpeg_packet_check (mpeg_demux_t *mpeg)
{
	if ((par_packet_max > 0) && (mpeg->packet.size > par_packet_max)) {
		return (1);
	}

	if (par_stream[mpeg->packet.sid] & PAR_STREAM_INVALID) {
		return (1);
	}

	return (0);
}

void mpeg_print_stats (mpeg_demux_t *mpeg, FILE *fp)
{
	unsigned i;

	fprintf (fp,
		"System headers: %lu\n"
		"Packs:          %lu\n"
		"Packets:        %lu\n"
		"End codes:      %lu\n"
		"Skipped:        %lu bytes\n",
		mpeg->shdr_cnt, mpeg->pack_cnt, mpeg->packet_cnt, mpeg->end_cnt,
		mpeg->skip_cnt
	);

	for (i = 0; i < 256; i++) {
		if (mpeg->streams[i].packet_cnt > 0) {
			fprintf (fp, "Stream %02x:      %lu packets / %llu bytes\n",
				i, mpeg->streams[i].packet_cnt, mpeg->streams[i].size
			);
		}
	}

	for (i = 0; i < 256; i++) {
		if (mpeg->substreams[i].packet_cnt > 0) {
			fprintf (fp, "Substream %02x:   %lu packets / %llu bytes\n",
				i, mpeg->substreams[i].packet_cnt, mpeg->substreams[i].size
			);
		}
	}

	fflush (fp);
}

int mpeg_copy (mpeg_demux_t *mpeg, FILE *fp, unsigned n)
{
	unsigned char buf[4096];
	unsigned      i, j;

	while (n > 0) {
		i = (n < 4096) ? n : 4096;

		j = mpegd_read (mpeg, buf, i);

		if (j > 0) {
			if (fwrite (buf, 1, j, fp) != j) {
				return (1);
			}
		}

		if (i != j) {
			return (1);
		}

		n -= i;
	}

	return (0);
}

int main (int argc, char **argv)
{
	int      argi;
	unsigned i;
	int      r;

	if (argc == 2) {
		if (str_isarg (argv[1], NULL, "--version")) {
			prt_version();
			return (0);
		}
		else if (str_isarg (argv[1], NULL, "--help")) {
			prt_help();
			return (0);
		}
	}

	for (i = 0; i < 256; i++) {
		par_stream[i] = 0;
		par_substream[i] = 0;
		par_stream_map[i] = i;
		par_substream_map[i] = i;
	}

	argi = 1;
	while (argi < argc) {
		if (str_isarg (argv[argi], "-c", "--scan")) {
			unsigned i;

			par_mode = PAR_MODE_SCAN;

			for (i = 0; i < 256; i++) {
				par_stream[i] |= PAR_STREAM_SELECT;
				par_substream[i] |= PAR_STREAM_SELECT;
			}
		}
		else if (str_isarg (argv[argi], "-l", "--list")) {
			par_mode = PAR_MODE_LIST;
		}
		else if (str_isarg (argv[argi], "-r", "--remux")) {
			par_mode = PAR_MODE_REMUX;
		}
		else if (str_isarg (argv[argi], "-d", "--demux")) {
			par_mode = PAR_MODE_DEMUX;
		}
		else if (str_isarg (argv[argi], "-s", "--stream")) {
			argi += 1;
			if (argi >= argc) {
				prt_err ("%s: missing stream id\n", argv[0]);
				return (1);
			}

			if (str_get_streams (argv[argi], par_stream, PAR_STREAM_SELECT)) {
				prt_err ("%s: bad stream id (%s)\n", argv[0], argv[argi]);
				return (1);
			}
		}
		else if (str_isarg (argv[argi], "-p", "--substream")) {
			argi += 1;
			if (argi >= argc) {
				prt_err ("%s: missing substream id\n", argv[0]);
				return (1);
			}

			if (str_get_streams (argv[argi], par_substream, PAR_STREAM_SELECT)) {
				prt_err ("%s: bad substream id (%s)\n", argv[0], argv[argi]);
				return (1);
			}
		}
		else if (str_isarg (argv[argi], "-i", "--invalid")) {
			argi += 1;
			if (argi >= argc) {
				prt_err ("%s: missing invalid stream id\n", argv[0]);
				return (1);
			}

			if (strcmp (argv[argi], "-") == 0) {
				for (i = 0; i < 256; i++) {
					if (par_stream[i] & PAR_STREAM_SELECT) {
						par_stream[i] &= ~PAR_STREAM_INVALID;
					}
					else {
						par_stream[i] |= PAR_STREAM_INVALID;
					}
				}
			}
			else {
				if (str_get_streams (argv[argi], par_stream, PAR_STREAM_INVALID)) {
					prt_err ("%s: bad stream id (%s)\n", argv[0], argv[argi]);
					return (1);
				}
			}
		}
		else if (str_isarg (argv[argi], "-b", "--base-name")) {
			argi += 1;
			if (argi >= argc) {
				prt_err ("%s: missing base name\n", argv[0]);
				return (1);
			}

			if (par_demux_name != NULL) {
				free (par_demux_name);
			}

			par_demux_name = str_clone (argv[argi]);
		}
		else if (str_isarg (argv[argi], "-m", "--packet-max-size")) {
			argi += 1;
			if (argi >= argc) {
				prt_err ("%s: missing maximum packet size\n", argv[0]);
				return (1);
			}

			par_packet_max = (unsigned) strtoul (argv[argi], NULL, 0);
		}
		else if (str_isarg (argv[argi], "-S", "--stream-map")) {
			unsigned id1, id2;

			if ((argi + 2) >= argc) {
				prt_err ("%s: missing stream id\n", argv[0]);
				return (1);
			}

			id1 = (unsigned) strtoul (argv[argi + 1], NULL, 0);
			id2 = (unsigned) strtoul (argv[argi + 2], NULL, 0);

			par_stream_map[id1 & 0xff] = id2 & 0xff;

			argi += 2;
		}
		else if (str_isarg (argv[argi], "-P", "--substream-map")) {
			unsigned id1, id2;

			if ((argi + 2) >= argc) {
				prt_err ("%s: missing substream id\n", argv[0]);
				return (1);
			}

			id1 = (unsigned) strtoul (argv[argi + 1], NULL, 0);
			id2 = (unsigned) strtoul (argv[argi + 2], NULL, 0);

			par_substream_map[id1 & 0xff] = id2 & 0xff;

			argi += 2;
		}
		else if (str_isarg (argv[argi], "-x", "--split")) {
			par_split = 1;
		}
		else if (str_isarg (argv[argi], "-h", "--no-system-headers")) {
			par_no_shdr = 1;
		}
		else if (str_isarg (argv[argi], "-k", "--no-packs")) {
			par_no_pack = 1;
		}
		else if (str_isarg (argv[argi], "-K", "--remux-skipped")) {
			par_remux_skipped = 1;
		}
		else if (str_isarg (argv[argi], "-t", "--no-packets")) {
			par_no_packet = 1;
		}
		else if (str_isarg (argv[argi], "-e", "--no-end")) {
			par_no_end = 1;
		}
		else if (str_isarg (argv[argi], "-E", "--empty-packs")) {
			par_empty_pack = 1;
		}
		else if (str_isarg (argv[argi], "-D", "--no-drop")) {
			par_drop = 0;
		}
		else if (str_isarg (argv[argi], "-F", "--first-pts")) {
			par_first_pts = 1;
		}
		else if (str_isarg (argv[argi], "-a", "--ac3")) {
			par_dvdac3 = 1;
		}
		else if (str_isarg (argv[argi], "-u", "--spu")) {
			par_dvdsub = 1;
		}
		else if ((argv[argi][0] != '-') || (argv[argi][1] == 0)) {
			if (par_inp == NULL) {
				if (strcmp (argv[argi], "-") == 0) {
					par_inp = stdin;
				}
				else {
					par_inp = fopen (argv[argi], "rb");
				}
				if (par_inp == NULL) {
					prt_err ("%s: can't open input file (%s)\n", argv[0], argv[argi]);
					return (1);
				}
			}
			else if (par_out == NULL) {
				if (strcmp (argv[argi], "-") == 0) {
					par_out = stdout;
				}
				else {
					par_out = fopen (argv[argi], "wb");
				}
				if (par_out == NULL) {
					prt_err ("%s: can't open output file (%s)\n", argv[0], argv[argi]);
					return (1);
				}
			}
			else {
				prt_err ("%s: too many files (%s)\n", argv[0], argv[argi]);
				return (1);
			}
		}
		else {
			prt_err ("%s: unknown parameter (%s)\n", argv[0], argv[argi]);
			return (1);
		}

		argi += 1;
	}


	if (par_inp == NULL) {
		par_inp = stdin;
	}

	if (par_out == NULL) {
		par_out = stdout;
	}

	switch (par_mode) {
	case PAR_MODE_SCAN:
		r = mpeg_scan (par_inp, par_out);
		break;

	case PAR_MODE_LIST:
		r = mpeg_list (par_inp, par_out);
		break;

	case PAR_MODE_REMUX:
		r = mpeg_remux (par_inp, par_out);
		break;

	case PAR_MODE_DEMUX:
		r = mpeg_demux (par_inp, par_out);
		break;

	default:
		r = 1;
		break;
	}

	if (r) {
		return (1);
	}

	return (0);
}
